dbt/include/clickhouse/macros/materializations/incremental.sql

```diff
@@ -234,3 +244,36 @@
     {% do adapter.drop_relation(new_data_relation) %}
     {{ drop_relation_if_exists(distributed_new_data_relation) }}
 {% endmacro %}
+
+{% macro clickhouse__incremental_insert_replace(existing_relation, intermediate_relation, partition_by) %}
+    {% set new_data_relation = existing_relation.incorporate(path={"identifier": model['name']
+       + '__dbt_new_data_' + invocation_id.replace('-', '_')}) %}
+    {{ drop_relation_if_exists(new_data_relation) }}
+    {% call statement('create_new_data_temp') -%}
+      {{ get_create_table_as_sql(False, new_data_relation, sql) }}
+    {%- endcall %}
+    {% call statement('main') -%}
+        create table {{ intermediate_relation }} as {{ existing_relation }}
+    {%- endcall %}
+    {% call statement('insert_new_data') -%}
+        insert into {{ intermediate_relation }} select * from {{ new_data_relation }}
+    {%- endcall %}
+    {% if execute %}
+      {% set select_changed_partitions %}
+          select distinct {{ partition_by|join(', ') }}
+          from {{ intermediate_relation }}
+      {% endset %}
+      {% set partitions_expressions = run_query(select_changed_partitions).rows %}
+    {% else %}
+      {% set partitions_expressions = [] %}
+    {% endif %}
+    {% for partition_expression in partitions_expressions %}
+      {% call statement('replace_partition_' ~ loop.index) %}
+          alter table {{ existing_relation }}
+          replace partition {{ partition_expression }}
```

Comment:
I am new to DBT and I need an advice for translating output of `run_query` into ClickHouse literal values. Currently `partitions_expressions` is an iterable of [`agate.Row`s](https://agate.readthedocs.io/en/latest/api/columns_and_rows.html#agate.Row). I need to translate `Row{someInt: 1, someStr: 'a', someDate: toDate('2023-11-01')}` to partition expression like `… partition (1, 'a', toDate('2023-11-01')`.

Additionally, we have `partition_by` list of columns names (without types) as an input.

I have several ideas to do it:
* Request target table's columns definitions. Pick the partitioning columns types. Use `CAST('<value-as-string-literal>' AS <target-type>)`.
* Implement translation as a method of `ClickHouseAdapter` and call `adapter.to_partition_expression(…)`. Though I am not sure about the best way to perform translation in adapter: I hope to leverage [`clickhouse-driver`](https://clickhouse-driver.readthedocs.io/en/latest/api.html#clickhouse_driver.Client.execute)'s functionality of translating `params` for this. However, the driver might not have a convenient public interface for the translation.
* Find out some SQL-only way to translate. I.e. compose an SQL query which will do the translation on ClickHouse level: smth like `with partition_expr as (select partitionBy offset {{ loop.index }} limit 1) alter table … replace partition partition_expr …`. This particular example does not work, but maybe there is a possible way.
* I am not sure if `clickhouse-driver`'s usage is discouraged in favour of HTTP interface. If so, a parameterized HTTP query might be an option, though looks the most complex option to me and still requires retrieving partitioning columns' types.

Please let me know if there is an easier dbt-style way to do it. Any advice regarding the suggested options or guidance towards using dbt-specific functions (like requesting a list of table's columns definitions) will be helpful.

I am doing my investigation of the approaches to choose the simplest one. Though I believe the first option is the most reasonable.

@genzgd @simpl1g @tanghyd given that you participated in the original issue #128, could you maybe share some thoughts please?

Reply:
See `get_partitions` macro in comment above, I believe it should solve this question

Reply:
Very helpful, thank you for your valuable feedback, Konstantin! 🤝 

I have implemented and tested you suggestion with selecting from `system.parts` and it works perfectly fine 👍